src/base/dof_map_constraints.C

```diff
@@ -1133,32 +1183,80 @@ public:
      * input function \p f gives the arbitrary solution.
      */
 
-    // Loop over all the variables we've been requested to project
-    for (const auto & var : dirichlet.variables)
-      {
-        const Variable & variable = dof_map.variable(var);
+    // Figure out which System the DirichletBoundary objects are
+    // defined for. Verify that all Variables are defined on the same
+    // System.
+    System * system = nullptr;
+    for (const auto & dirichlet : dirichlets)
+      for (const auto & var : dirichlet->variables)
+        {
+          const Variable & variable = dof_map.variable(var);
+          auto current_system = variable.system();
 
-        const FEType & fe_type = variable.type();
+          if (!system)
+            system = current_system;
+          else if (current_system != system)
+            libmesh_error_msg("All variables should be defined on the same System");
+        }
 
-        if (fe_type.family == SCALAR)
-          continue;
+    // If we found no System it could be because there are no
+    // DirichletBoundary objects or there are DirichletBoundaries but
+    // they have no Variables defined on them? This seems like a
+    // strange situation but it would also imply there is nothing to
+    // be done, so let's just return in that case.
+    if (!system)
+      return;
```

Comment:
This, on the other hand, seems a bit far in the other direction; we can test for those cases from within library code so if we find ourselves here anyway I'd rather libmesh_assert(system) and figure out what went wrong.

Reply:
OK, I will change this.